### ISSUE TITLE
Verify accessibility

### ISSUE THREAD

**Issue by aforcier** [wordpress-mobile/WordPress-Editor-Android#183]
Placeholder issue - test the app for accessibility.

**Comment by bummytime** [comment]
:+1:

**Comment by aforcier** [comment]
I did some accessibility testing for the editor. Following the [guidelines](http://developer.android.com/tools/testing/testing_accessibility.html#how-to), I enabled _TalkBack_ and _Explore by Touch_ and navigated the app (I tried both our example app and WPAndroid).

Here are my findings:
- Navigating by swipe (this traverses all the elements on the page) goes through each item in the `ActionBar`, then to the `WebView`, then to the title field (inside the `WebView`), then to the content field (also inside the `WebView`), then to the format buttons at the bottom.
- When highlighting the title and content fields, the contents are read out to you. If they're empty, the placeholders are read out ("Post title", "Share your story here...")
- Our format bar buttons don't have accessibility labels, so they're all read out as unlabeled checkboxes
- When the `WebView` overall is highlighted, the audio description is "ZSSRichTextEditor". The culprit is probably the title of the editor HTML file:

https://github.com/wordpress-mobile/WordPress-Editor-Android/blob/ca9c4eda2f2f13ad24790cfa1046cc8c2e5a024e/libs/editor-common/assets/android-editor.html#L4
- Attempting to edit with the on-screen keyboard is abysmal when _Explore by Touch_ is enabled (navigating the screen through gestures). It looks like `contenteditable` just isn't properly supported by the Android `WebView` for accessibility. Though you can have the contents read to you (and configure whether to read character-by-character, word-by-word, and so on), the cursor doesn't move. And repositioning the cursor by touch doesn't work well at all. This could be because I'm a novice at using accessibility services, but I doubt it: when using a native `EditText` (e.g. the legacy editor, or our own editor's HTML mode) everything behaves much more smoothly.
- The experience is much better when using a hardware keyboard, since you can move the cursor around. I'm not sure how common it is for users using _Explore by Touch_ to want to write text with the software keyboard - I haven't been able to find any useful information about that yet.

I'm not sure how we can fix the software keyboard issue in _Explore by Touch_ mode yet. I think for now we mainly need to fix the accessibility issues we can, and have a good flow for transitioning users to legacy or to HTML mode if the visual editor isn't working for their accessibility setup.

In summary, the main things we should do are:
- [x] Label the format bar buttons
- [x] Change the title of the HTML file to something more appropriate ("Visual Editor" perhaps)
- [x] The post and image settings screens both need accessibility labels for the various fields
- [ ] Help users of the new editor whose setup isn't working with the visual editor (right now this looks like _Explore by Touch_ + software keyboard, but there are likely other problematic cases). The overall page navigation works, so maybe we can add a way to change to the legacy editor, or default to HTML mode? Otherwise, some sort of prompt, something like what @astralbodies suggested here: https://github.com/wordpress-mobile/WordPress-Editor-iOS/issues/608#issuecomment-115333537
- [ ] Related to the above, we should make the switch between visual/HTML mode more obvious. Maybe a toast ("Switched to visual mode", "Switched to HTML mode"), those seem to be read out automatically (this also came up in the iOS editor issue above).

There are some more extensive problems though, beyond our editor. WPAndroid itself obviously needs some accessibility work. Most relevant to us is that it seems that when one of the other fragments is shown (post settings, image settings), the editor fragment is read out instead. There are also some smaller things I noticed, like the 'new post' FAB is unlabelled, as well as the masterbar items.

Any thoughts @bummytime, @meremagee?

**Comment by maxme** [comment]
> When the WebView overall is highlighted, the audio description is "ZSSRichTextEditor". The culprit is probably the title of the editor HTML file:

Confirmed, changing the HTML title fixes the problem.

> When highlighting the title and content fields, the contents are read out to you. If they're empty, the placeholders are read out ("Post title", "Share your story here...")

IMO, it's ok to keep these placeholders - distinguishable enough, and I guess it's part of the "Talkback learning process".

>  format bar buttons

Should be changed in the legacy editor too (:fire: ToggleButton :fire:) 

> Related to the above, we should make the switch between visual/HTML mode more obvious. Maybe a toast ("Switched to visual mode", "Switched to HTML mode"), those seem to be read out automatically (this also came up in the iOS editor issue above).

Instead of showing a Toast, we might explore Accessibility events: https://developer.android.com/reference/android/view/accessibility/AccessibilityEvent.html

> WPAndroid itself obviously needs some accessibility work

Yes ;)

**Comment by aforcier** [comment]
> IMO, it's ok to keep this placeholders - distinguishable enough, and I guess it's part of the "Talkback learning process".

Agreed, I didn't make it clear in my original post but I considered the placeholders being read out to be a positive :grinning: 

> Instead of showing a Toast, we might explore Accessibility events: https://developer.android.com/reference/android/view/accessibility/AccessibilityEvent.html

:+1:

**Comment by maxme** [comment]
> Instead of showing a Toast, we might explore Accessibility events: https://developer.android.com/reference/android/view/accessibility/AccessibilityEvent.html

We can do something like: `view.announceForAccessibility("HTML mode enabled");` or send `TYPE_ANNOUNCEMENT` events. But these messages are uninterruptible and break the normal TTS flow. IMO, we should not use that, if a user activates the "HTML Mode switch button", he must expect the html mode to turn on.
